src/api/products.ts

```diff
@@ -0,0 +1,12 @@
+import client from './client';
+
+export type ResponseReturn<T> = {
+  ok: boolean;
+  data?: T;
+};
```

Comment:
```suggestion
export type ResponseReturn<T> =
  | {
      ok: true;
      data: T;
    }
  | {
      ok: false;
      data: null;
    };
```
타입 정의하실 때 각 상태별로 타입을 좁혀둘 수 있으면 이렇게 좁혀두는걸 추천드려요~

Reply:
@EungyuCho 

- 질문인데요, 타입을 좁히는 이유는 무엇인가요?
- 상황별로 데이터를 확인하기 더 좋아서 그런 이유일까요?

Reply:
맞아요 성공했을때 케이스와 실패하는 케이스를 나눠서 풀 수 있어서 코멘트를 드렸어요.
두가지를 한번에 풀다보면 각 속성마다 optional한 데이터를 핸들링해줘야하기도 하고 실수할 여지도 늘어나는 것 같아서요!

```typescript
// 성공과 실패의 케이스를 나눈 경우
type Product = {
  name: string;
  price: number;
  quantity: number;
};

type ProductCard = {
  product: Product | null;
  totalPrice: number;
};

function getProductCardData<T>(input: ResponseReturn<Product>): ProductCard {
  if (!input.ok) {
    return {
      product: null,
      totalPrice: 0,
    };
  }

  return {
    product: input.data,
    totalPrice: input.data.price * input.data.quantity,
  };
}

// 한번에 푸는 경우
function getProductCardData<T>(input: ResponseReturn<Product>): ProductCard {
  return {
    product: input.data,
    totalPrice: (input?.data?.price ?? 0) * (input?.data?.quantity ?? 0),
  };
}

```